MusicNotationKit/MusicNotationKit/Tuplet.swift

```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
+                && (!isFlatIndex(afterLast, sameTupletAs: lastFlatIndex) || lastFlatIndex.count == 1)
+        }
+    }
+
+    /**
+     Returns a modified `NoteCollection` with the tie states correct for preserving the original tie states of the notes
+     being replaced. The range may also include notes with tie states that represent an invalid state for modification.
+     */
+    private func preserveTieStateForReplacement(in range: CountableClosedRange<Int>,
+                                                with newCollection: NoteCollection) throws -> NoteCollection {
+        return try preserveTieStateForReplacement(in: range, with: [newCollection])[0]
+    }
+
+    private func preserveTieStateForReplacement(in range: CountableClosedRange<Int>,
+                                                with newCollections: [NoteCollection]) throws -> [NoteCollection] {
+        var modifiedCollections = newCollections
+        let lastIndex = modifiedCollections.count - 1
+
+        func modifyCollections(atFirst first: Bool, with note: Note) throws {
+            let index = first ? 0 : lastIndex
+            if var tuplet = modifiedCollections[index] as? Tuplet {
+                try tuplet.replaceNote(at: first ? 0 : tuplet.flatIndexes.count - 1, with: note)
+                modifiedCollections[index] = tuplet
+            } else {
+                modifiedCollections[index] = note
+            }
+        }
+
+        func modifyState(forTie originalTie: Tie?) throws {
+            switch originalTie {
+            case .begin?:
+                var lastNote = newCollections[lastIndex].last
+                try lastNote.modifyTie(.begin)
+                try modifyCollections(atFirst: false, with: lastNote)
+            case .end?:
+                var firstNote = newCollections[0].first
+                try firstNote.modifyTie(.end)
+                try modifyCollections(atFirst: true, with: firstNote)
+            case .beginAndEnd?:
+                if newCollections.count == 1 && newCollections[0].noteCount == 1 {
+                    var onlyNote = try newCollections[0].note(at: 0)
+                    try onlyNote.modifyTie(.beginAndEnd)
+                    modifiedCollections[0] = onlyNote
+                } else {
+                    var firstNote = newCollections[0].first
+                    try firstNote.modifyTie(.end)
+                    var lastNote = newCollections[lastIndex].last
+                    try lastNote.modifyTie(.begin)
+                    try modifyCollections(atFirst: true, with: firstNote)
+                    try modifyCollections(atFirst: false, with: lastNote)
+                }
+            case nil:
+                break
+            }
+        }
+
+        if range.count == 1 {
+            let originalTie = try note(at: range.lowerBound).tie
+            try modifyState(forTie: originalTie)
+        } else {
```

Comment:
`return modifiedCollections` instead of `else`? Should we still try to return early in all cases?